### PR TITLE
crm-19150 Preserve contribution when associated membership is deleted…

### PR DESCRIPTION
**Pull request by joseltorres**
Currently when a membership is deleted, if there is an associated contribution then the contribution also gets deleted. This adds the option to preserve the contrition. I suggest this be completed in two steps: 1) The ability the preserver the contribution through code (API and BAO) and 2) UI change to allow front-end users to check a box to preserve the associated contribution. This is step 1.

https://issues.civicrm.org/jira/browse/CRM-19150

### CODE REVIEW

File: api/v3/Membership.php [hunk 1]
```diff
@@ -42,7 +59,17 @@
  *   API result array.
  */
 function civicrm_api3_membership_delete($params) {
-  return _civicrm_api3_basic_delete(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (isset($params['preserve_contribution'])) {
+    if (CRM_Member_BAO_Membership::del($params['id'], $params['preserve_contribution'])) {
+      return civicrm_api3_create_success(TRUE, $params);
+    }
+    else {
+      return civicrm_api3_create_error('Could not delete membership');
```
Comment (seamuslee001) [inline api/v3/Membership.php]: Hi @joseltorres Hi Jose, I think the better thing to do here would be to throw a exception but not sure anythoughts @eileenmcnaughton

**Comment by seamuslee001** [comment]
@joseltorres This is looking good I do like it,  A couple of things 1. Can do you do the following whilst on the branch for this work.
git fetch <repository alias for the core repository> 
git rebase -i <alias for the core repository/master> make sure you say to keep all your commits
git push <alias for your repo> master -f 

2nd thing i would say is that in all the BAO functions i think we should use preserve_contribution as a booleen rather than integer of 0. So have it TRJUE or FALSE given that is how the code is looking anyway

**Comment by seamuslee001** [comment]
@joseltorres can you please run the following chmod 0644 CRM/Member/BAO/Membership.php api/v3/Membership.php I have noticed you have changed the permissions on these 2 files but the permissions should be 644 not 755 (i.e. owner readwrite and group and others read not owner readwrite and execute and read and execute for group and others

File: api/v3/Membership.php [hunk 2]
```diff
@@ -42,7 +59,17 @@
  *   API result array.
  */
 function civicrm_api3_membership_delete($params) {
-  return _civicrm_api3_basic_delete(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (isset($params['preserve_contribution'])) {
+    if (CRM_Member_BAO_Membership::del($params['id'], $params['preserve_contribution'])) {
+      return civicrm_api3_create_success(TRUE, $params);
```
Comment (seamuslee001) [inline api/v3/Membership.php]: hi @joseltorres looking over this again, I think what might be the best idea here is to actually remove the isset and just have _civicrm_api3_basic_delete(_civicrm_api3_get_BAO(__FUNCTION__), $params); Then in the CRM_Membership_BAO_Membership::del() function have $preserveContrib = isset($params['preserve_contribution']) ? $params['preserve_contribution'] : FALSE; 

That way then we aren't doing anything strange at the API level and getting it all to be handled at the BAO level

File: api/v3/Membership.php (hunk 1 above)
Comment (eileenmcnaughton) [inline api/v3/Membership.php]: Yes - definitely we prefer to throw exceptions

File: api/v3/Membership.php (hunk 2 above)
Comment (eileenmcnaughton) [inline api/v3/Membership.php]: Less at the api level is better

File: api/v3/Membership.php [hunk 3]
```diff
@@ -33,6 +33,23 @@
...
+ * The metadata is used for setting defaults, documentation & validation.
+ *
+ * @param array $params
+ *   Array of parameters determined by getfields.
+ */
+function _civicrm_api3_membership_delete_spec(&$params) {
+  $params['preserve_contribution'] = array(
+    'api.required' => 0,
+    'title' => 'Preserve Contribution',
+    'description' => 'By default this is 0, or 0 if not set. Set to 1 to preserve the associated contribution record when membership is deleted.',
```
Comment (eileenmcnaughton) [inline api/v3/Membership.php]: Actually I'm really really happy to change the default to preserve by default. In fact at the API level I'm OK with us not supporting delete the contribution as well at all.

We'd have to announce it but I feel the existing behaviour is dangerous

File: api/v3/Membership.php (hunk 2 above)
Comment (eileenmcnaughton) [inline api/v3/Membership.php]: Per my comments - I'm actually OK with us not supporting this option from the api level at all. I feel like if people using the api want to delete payment information they can call contribution.delete. I expect more people are accidentally deleting wanted data than enjoying the functionality they are getting at the moment

File: api/v3/Membership.php (hunk 3 above)
Comment (seamuslee001) [inline api/v3/Membership.php]: I agree but i think its best done through the spec rather than the delete function IMO

Comment (eileenmcnaughton) [inline api/v3/Membership.php]: I guess I was saying I'd be happy to go further - ie. I'm saying the default should be preserve not delete - but at that point people who want to mess up peoples' databases have to change their code anyway - at which point I don't know if it's worth supporting delete-contribution-via-membership-api at all - they can do that in their code in the very unusual case where they really want to

**Comment by monishdeb** [comment]
Oops I haven't seen this PR and created another one https://github.com/civicrm/civicrm-core/pull/8821

File: api/v3/Membership.php (hunk 3 above)
Comment (joseltorres) [inline api/v3/Membership.php]: My concern with @eileenmcnaughton suggestion is that it changes the current behavior. If someone has existing code that deletes a membership with an associated contribution, they will be expecting that contribution to also be deleted. I can see preserving the contribution being a desired behavior, but only if this were a new API or BAO action. My goal was to add this improvement without changing the way any existing calls this this API or BAO behave.

File: api/v3/Membership.php (hunk 2 above)
Comment (seamuslee001) [inline api/v3/Membership.php]: Following comments from Mooish on a separate PR I am satisfied with this approach

**Comment by seamuslee001** [comment]
@joseltorres can you close this in favour of https://github.com/civicrm/civicrm-core/pull/8825

**Comment by joseltorres** [comment]
closing in favor of https://github.com/civicrm/civicrm-core/pull/8825

**Pull request closed without merging**